### PR TITLE
- added code from internal repository

### PR DESCRIPTION
**Pull request by mobilunity-user**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
+{
+  "name": "ravendb-node",
```
Comment (gregolsky) [inline package.json]: Let's change the name of the package to `ravendb-client`.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    "url": "git@git.devs.domerp.com:frontend-projects/ravendb-node.git"
+  },
+  "keywords": [
+    "ravendb",
+    "typescript",
+    "node",
+    "nodejs",
+    "nosql"
+  ],
+  "author": "Hybernating Rhinos",
```
Comment (gregolsky) [inline package.json]: A typo here. It's *Hibernating Rhinos LTD*

File: package.json [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  "scripts": {
+    "_prepublish": "npm run bundle",
+    "quick-build": "tsc --pretty",
+    "bundle": "node scripts/build-bundle.js",
+    "pretest": "node scripts/build-tests.js",
+    "test": "mocha"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git@git.devs.domerp.com:frontend-projects/ravendb-node.git"
```
Comment (gregolsky) [inline package.json]: Please change the url to https://github.com/ravendb/ravendb-nodejs-client

File: package.json [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  },
+  "keywords": [
+    "ravendb",
+    "typescript",
+    "node",
+    "nodejs",
+    "nosql"
+  ],
+  "author": "Hybernating Rhinos",
+  "license": "ISC",
```
Comment (gregolsky) [inline package.json]: Please set the `license` field to [match the one set on the repository](https://github.com/ravendb/ravendb-nodejs-client/blob/v4.0/LICENSE).

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
+{
+  "version": "2.2",
+  "compilerOptions": {
+    "target": "es5",
```
Comment (gregolsky) [inline tsconfig.json]: target mismatch with https://github.com/ravendb/ravendb-nodejs-client/pull/1/commits/b0919fb7aa07a8f95cf7c3cdf0ebbe3684e155be#diff-601a9423323a9d2a4fd0c9b51b5ae4caR42

File: src/Documents/DocumentStore.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
+import {IDocument} from './IDocument';
+import {IDocumentStore} from './IDocumentStore';
+import {IDocumentSession} from "./Session/IDocumentSession";
+import {DocumentSession} from "./Session/DocumentSession";
+import {RequestExecutor} from '../Http/RequestExecutor';
+import {DocumentConventions} from './Conventions/DocumentConventions';
+
```
Comment (gregolsky) [inline src/Documents/DocumentStore.ts:7]: Let's use the following style conventions for Typescript code in this project:

Use PascalCase for class and interface names.
Use PascalCase for enum values.
Use camelCase for function names.
Use camelCase for property names and local variables.
Use camelCase for argument names.
Use "_" as a prefix for `private` properties (not for `protected` ones).
Use camelCase for file / module names;

File: src/Documents/DocumentStore.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+export class DocumentStore implements IDocumentStore
+{
+  protected Url: string;
+  protected Database: string;
+  protected ApiKey?: string;
+  protected _RequestExecutor: RequestExecutor;
+  protected _Conventions: DocumentConventions; 
+
+  public get RequestExecutor(): RequestExecutor
```
Comment (gregolsky) [inline src/Documents/DocumentStore.ts]: Please keep one bracket opening/closing convention - opening brackets on line end and closing ones in new lines.

File: scripts/build-bundle.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+        const filtered = lines
+            .filter(line => !line.startsWith('import'))
+            .map(line => line.replace(/export /, ''))
+            .join('\n');
+
+        return `/* file: ${file} */\n${filtered}`;
+    }).join('\n');
+
+fs.writeFileSync(bundle, allContents, { encoding: 'utf8', flag: 'a' });
```
Comment (gregolsky) [inline scripts/build-bundle.js]: Are we going to convert this to `gulp` tasks ?

**Review by gregolsky — CHANGES_REQUESTED** [review]

File: src/Documents/Document.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import {IDocument} from './IDocument';
+
+export class Document implements IDocument {
+  public entityName(): string {
+      return this.constructor.name;
+  }
+
```
Comment (ayende) [inline src/Documents/Document.ts:7]: Would we need serialize / deserialize? We are working on JS, after all. We can just do `JSON.serialize` and vice versa and get the same behavior, no?

Comment (gregolsky) [reply src/Documents/Document.ts:7]: We probably do not need a class nor interface for `Document` at all. `Raven.Client` (C#) does not have a concept of document. It just returns objects/entities. We could just use `JSON.parse()` on the document body from server's response and `JSON.serialize()` if we need a string representation.

Comment (ayende) [reply src/Documents/Document.ts:7]: Agree, we should be able to accept any JS object here.

Comment (gregolsky) [reply src/Documents/Document.ts:7]: @a-serdukov-mobilunity please adjust the code based on our above comments. If you have any questions concerning this change let's talk tomorrow.

**Comment by gregolsky** [comment]
@a-serdukov-mobilunity Please fix the conflicts with LICENSE file. Thanks.

**Comment by ayende** [comment]
We need to pull this already, so there is _something_ that we can start with

**Comment by gregolsky** [comment]
@ayende OK, I've rebased changes and created PR #2 . We should close this one.

File: src/Documents/Document.ts (hunk 9 above)
Comment (mobilunity-user) [reply src/Documents/Document.ts:7]: how i should serialize "any js object" ? i need some predefined methods like fromJSON/toJSON. 
your Python client accepts any objects, it also support linked documents where the property is a another class

also, a lot of mongo clients has own models classes with validation, relations etc (mongoose for example)
i would use IDocument/Document based on Backbone Model - they allows more features that simple js object literal

Comment (ayende) [reply src/Documents/Document.ts:7]: `JSON.serialize` should handle it, no?
And validation is something that you do outside of the object, using listeners, for example.

**Pull request closed without merging**